bindings/js/api.js

```diff
@@ -1,76 +1,207 @@
 // jshint esversion: 6
 
-var fs = require('fs');
-var path = require('path');
-var loader = require(path.resolve(__dirname, './loader.js'));
-var shell = require(path.resolve(__dirname, './shell.js'));
-var api_json = require(path.resolve(__dirname, './api.json'));
+if (typeof module !== 'undefined') {
+  var fs = require('fs');
+  var path = require('path');
+  var loader = require(path.resolve(__dirname, './loader.js'));
+  var shell = require(path.resolve(__dirname, './shell.js'));
+  var api_promise = Promise.resolve(require(path.resolve(__dirname, './api.json')));
+  var layouts_promise = Promise.resolve(require(path.resolve(__dirname, './layouts.json')));
+  var modules_promise = Promise.resolve(shell.my_modules.map(m => {
+    var source = fs.readFileSync(path.resolve(__dirname, './' + m + ".wasm"));
+    return new Uint8Array(source);
+  }));
+
+} else {
+  var loader = this;
+  var shell = this;
+  var api_promise = fetch("api.json").then(r => r.json());
+  var layouts_promise = fetch("layouts.json").then(r => r.json());
+  var modules_promise = Promise.all(my_modules.map(m => fetch(m + ".wasm")))
+    .then(response => Promise.all(response.map(r => r.arrayBuffer())));
+}
+
+// Uncomment for debug
+loader.setMyPrint((x) => {});
+
+// HELPERS FOR SIZE FIELDS
+// -----------------------
+
+// Grammar of size fields:
+//
+// f ::=
+//   var
+//   var<OP>n
+//   M.f(var)
+//
+// var: variable, one of the length variables, may be ghost
+// n:   integer constant
+// <OP>: +, -, * or /
+// M.f: function, referred to by its name in the high-level API, e.g.
+//      EverCrypt_Hash.hash_len (and not EverCrypt_Hash_Incremental_hash_len)
+
+var parseOp = (arg, op) => {
+  let [ var_, const_ ] = arg.split(op);
+  return [ "Op", var_, op, const_ ];
+};
+
+var parseApp = (arg) => {
+  let i = arg.indexOf("(");
+  let [ m, f ] = arg.substr(0, i).split(".");
+  let x = arg.substr(i+1, arg.length - 1 - i - 1);
```

Comment:
Why isn't `parseInt` called here once instead of multiple times later in `evalSize` and `invertSize`?
Also, shouldn't `x` be checked to be a number (and throw an error if that's not the case)?

Reply:
see https://github.com/project-everest/hacl-star/blob/protz_wasm_bignum/bindings/js/api.json#L3059 for an example of the `App` case... it might be `f(x)` where both `f` and `x` are variables, not concrete numbers